### DISCUSSION TITLE
Rewrite 'assert' statements as if/raise?

### DISCUSSION THREAD

**Discussion by adamchainz** [encode/django-rest-framework#7832] [Ideas & Suggestions]
## Checklist

- [x] Raised initially in discussion on [Django ticket #32508](https://code.djangoproject.com/ticket/32508)
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [x] I have reduced the issue to the simplest possible case.

Python's `-O` flag, enables "optimized mode" which, among other things, disables the `assert` statement. A naive user may activate it to try speed up their application. DRF has many uses of `assert` in its library code (not tests) to guard against bad conditions. Rewriting those `assert` statements as an `if` check and a `raise` of an appropriate exception would allow it to continue functioning in "optimized mode".

**Comment by lovelydinosaur** [comment]
Thanks Adam!

I'd *probably* prefer not. (Tho I'm open to being persuaded here.) If a user is turning off assertions, then, well, let 'em turn off assertions.

But how about this, shall we get this discussion to a more concrete point first, by taking a single specific example of one of the assert statements in question, to help figure out what the benefits/trade-offs would look like in a particular concrete case?

**Reply by adamchainz** [reply]
Fair enough. I'm not really that fussed about this, I just thought it would be nice to prevent the worst footgun effects of `-O` and to follow Django. I checked some `assert`s in DRF and can't point to a specific case where it is definitely warranted.
